File: 0224-basic-calculator/0224-basic-calculator.py
class Solution:
    def calculate(self, s: str) -> int:
        stack = []
        number = 0
        res = 0
        sign = 1

        for i in s:
            if i.isdigit():
                number = (number * 10) + int(i)
            elif i == '+':
                res += (number * sign)
                number = 0
                sign = 1
            elif i == '-':
                res += (number * sign)
                number = 0
                sign = -1
            elif i == '(':
                stack.append(res)
                stack.append(sign)
                res = 0
                sign = 1
            elif i == ')':
                res += sign * number
                res *= stack.pop()
                res += stack.pop()
                number = 0
            
        return res + number * sign

File: 0224-basic-calculator/0224-basic-calculator.java
class Solution {
    public int calculate(String s) {
        Stack<Integer> stack = new Stack<Integer>();
        int result = 0, num = 0, sign = 1;
        
        for(int i=0; i<s.length(); i++){
            char c = s.charAt(i);
            
            if(Character.isDigit(c)){
                num = num*10 + (c-'0');
                System.out.println(num);
            }
            else if(c == '+'){
                result += num*sign;
                num = 0;
                sign = 1;
            }
            else if(c == '-'){
                result += num*sign;
                num = 0;
                sign = -1;
            }
            else if(c == '('){
                stack.push(result);
                stack.push(sign);
                result = 0;
                num = 0;
                sign = 1;
            }
            else if(c == ')'){
                
                result += num*sign;
                num = 0;
                result = result*stack.pop();
                result = result + stack.pop();
            }
        }
        result += num*sign;
        
        return result;
        
        
        
    }
    
    
}